### ISSUE TITLE
Suggestions, ideas and wishlist for Python interoperation

### ISSUE THREAD

**Issue by prabhuramachandran** [chapel-lang/chapel#16049]
This is really not an issue per se but as the title says, these are ideas and suggestions.  This is also perhaps not the best informed and my suggestions are with respect to what I read on the current documentation (https://chapel-lang.org/docs/).  This "issue" is posted as a result of some discussions on a SciPy 2020 slack chat with @bradcray and others.  Apologies for the length of this issue.

My main comments are in the area of Python interoperability.  I guess it is easiest if I start off with what I would like to see as a long time Python user, and contributor who has looked at chapel multiple times in the past.

I think the biggest strength with chapel is the ability to handle both single-node (multi-core), multi-node, and GPU support (I assume that it would be just as seamless to work with a GPU when it is supported) in a with language which is designed for parallel computing at scale.  This is an area where the only other contender I have seen is perhaps charm4py  (https://charm4py.readthedocs.io/en/latest/).  I will not go into a comparison here but given this, I think having the following features is super useful.

I would like to be able to use chapel from Python but in a more Pythonic fashion than exposed currently.  I would like to be able to write say a set of functions or expose some objects implemented in Chapel and pass numpy arrays back and forth to these.  A nice way to think of this is how say PyOpenCL or PyCUDA works, you can write GPU kernels, pass them arrays but you drive everything from Python.  There are explicit notions of sending data back and forth from the host to the device and vice-versa but there is a simple way for a Python user to work entirely from the Python world and compose OpenCL/CUDA programs.  I work on a very simple project that abstracts these between OpenMP/CUDA and OpenCL here (https://github.com/pypr/compyle) but that is a very young and simple project. There are many more examples but I think you get the idea.  

Currently, the way the documentation suggests Python interoperation works is more "monolithic" in that you explicitly initialize the chapel runtime with the `chpl_setup()` call and then the `chpl_cleanup()`.   Furthermore it seems one cannot have multiple libraries.  Ideally, what one would like is to build a set of chapel functionality and have the ability to call them in a user-defined manner.  The chpl_cleanup also exits which is not desirable either.  Perhaps a better approach would be to do what say mpi4py (https://mpi4py.readthedocs.io) does, where they initialize mpi at one point and users are either explicitly supposed to do so or there is an implicit initialization and the finalization is done automatically by registering a callback with the `atexit` module.

Something like the following would be nice.  Consider these files:
```
a.chpl
b.chpl
driver.py
```
Now in `driver.py`:

```
import numpy as np
import chapel
chapel.init() # this can be implicit in the import if desired.  This loads the runtime.

chapel.config() # could set up the configurable pieces, number of locales or whatever.
import a
import b
x = np.linspace(0, 10, 100000)

print(a.parallel_sum(x))
b.foo(x)
```
As you can see, this would be really quite convenient.

Tools like cppimport (https://github.com/tbenthompson/cppimport) allow folks to seamlessly call C++ from Python with very nice support for numpy arrays as well.  It might be worth looking at the kind of features and configurability that is exposed there.

The import hook is not necessary for a first cut implementation as this can always be added later but at least the possibility of doing things like:
```
import chapel
module = chapel.build(
'''
# Insert chapel code here.
'''
)

module.f(1.0)
```
This could be used in many interesting ways.  Again, none of this is new, this is pretty standard in many Python libraries but I think having these basic pieces in a core Chapel provided API would be really fantastic to improve the Python interoperability.  I am not saying that the `build` from source function is critical, it is actually pretty trivial to implement with a subprocess call currently but what is more important is that the chapel Python API is clear and exposes these kinds of things.  

I think improvements like this would allow more folks to start experimenting with chapel from the Python world.  Thanks for reading!

**Comment by lydia-duncan** [comment]
Hello!  Thanks for your interest in Chapel and especially in Python interoperability.

> Furthermore it seems one cannot have multiple libraries.

We have heard of situations where this worked, but it may be platform dependent.  This is definitely something we'd like to support more solidly, but the main barrier to doing so is that we would have to significantly rearrange our runtime and generated code, which makes it hard to prioritize.  All that to say, we hear you :)

> The chpl_cleanup also exits which is not desirable either. 

Note that #9290 talks about this issue generally for our interoperability story.

> Tools like cppimport (https://github.com/tbenthompson/cppimport) allow folks to seamlessly call C++ from Python with very nice support for numpy arrays as well. It might be worth looking at the kind of features and configurability that is exposed there.

I'll definitely take a look at that!  When considering tools for how to support Python interoperability, we checked [this list](https://wiki.python.org/moin/IntegratingPythonWithOtherLanguages) and it looks like that wasn't present on it, so it's definitely new to us.  From a policy standpoint, we're unlikely to switch away from a tool that is as well known in the Python community as Cython unless we are certain the tool we would be using is equally as well regarded and supported, but that doesn't preclude us from learning lessons or strategies about interoperability from it!

The `chapel.build` call you describe is actually a bit reminiscent of our original interoperability strategy, [PyChapel](https://github.com/chapel-lang/pychapel) (now deprecated), though PyChapel was quite a bit more heavy-weight.  It's something that we haven't thought about as much recently, as it seems very different from the other types of interoperability we're used to.  That's not to say it isn't a valuable strategy, just that our impression is that it is more Python-specific and perhaps limited to a subset of that community.  It's not off the table, and is definitely something we'll have to consider more

**Comment by prabhuramachandran** [comment]
Hi @lydia-duncan -- thanks for your quick response!  My suggestions are indeed Python specific. 

> > Tools like cppimport (https://github.com/tbenthompson/cppimport) allow folks to seamlessly call C++ from Python with very nice support for numpy arrays as well. It might be worth looking at the kind of features and configurability that is exposed there.
> 
> I'll definitely take a look at that! When considering tools for how to support Python interoperability, we checked [this list](https://wiki.python.org/moin/IntegratingPythonWithOtherLanguages) and it looks like that wasn't present on it, so it's definitely new to us. From a policy standpoint, we're unlikely to switch away from a tool that is as well known in the Python community as Cython unless we are certain the tool we would be using is equally as well regarded and supported, but that doesn't preclude us from learning lessons or strategies about interoperability from it!

Cython also has support for something like this via pyximport (see here: https://cython.readthedocs.io/en/latest/src/userguide/source_files_and_compilation.html#compiling-with-pyximport) , the issue is not so much the ability to directly import the Chapel-Python-extension module. That is easy to create with an importhook as both pyximport and cppimport do.  What is perhaps more important is that there be a separation of the runtime initialization and loadup from the finalization.  mpi4py does this.


> 
> The `chapel.build` call you describe is actually a bit reminiscent of our original interoperability strategy, [PyChapel](https://github.com/chapel-lang/pychapel) (now deprecated), though PyChapel was quite a bit more heavy-weight. It's something that we haven't thought about as much recently, as it seems very different from the other types of interoperability we're used to. That's not to say it isn't a valuable strategy, just that our impression is that it is more Python-specific and perhaps limited to a subset of that community. It's not off the table, and is definitely something we'll have to consider more

Yes, indeed, this is highly Python specific.   I think the separation of the initialization and finalization is more important and once done, the rest is easy to do from pure Python and if I have the time I'd be happy to help but I'm a little hard-pressed right now but this would make it much easier for interoperability.  

Is there a timeline for when numpy support will be available?  Thanks!

**Comment by lydia-duncan** [comment]
> Cython also has support for something like this via pyximport (see here: https://cython.readthedocs.io/en/latest/src/userguide/source_files_and_compilation.html#compiling-with-pyximport) , the issue is not so much the ability to directly import the Chapel-Python-extension module. That is easy to create with an importhook as both pyximport and cppimport do.

Ohhh, interesting.  Our library compilation step ends up generating the .pyx files from the .chpl file, so we'd need more control over what occurs in the import step, to get it to trigger that compilation in addition to the setup.py related call.  It may be that we can provide an implementation that mimics the important portions of Cython's, though, using the importhook you describe.  Definitely worth looking into.

> Is there a timeline for when numpy support will be available?

There's not an explicit timeline.  We have some numpy support already (in that specific numpy arrays can be passed to Chapel functions today) but it isn't *full* numpy support.  We do have it on our priority list, but may not get back to extending it this release (and I expect full numpy support to take longer than that), it all depends on the other priorities and developer resources.

**Comment by prabhuramachandran** [comment]
> > Cython also has support for something like this via pyximport (see here: https://cython.readthedocs.io/en/latest/src/userguide/source_files_and_compilation.html#compiling-with-pyximport) , the issue is not so much the ability to directly import the Chapel-Python-extension module. That is easy to create with an importhook as both pyximport and cppimport do.
> 
> Ohhh, interesting. Our library compilation step ends up generating the .pyx files from the .chpl file, so we'd need more control over what occurs in the import step, to get it to trigger that compilation in addition to the setup.py related call. It may be that we can provide an implementation that mimics the important portions of Cython's, though, using the importhook you describe. Definitely worth looking into.

Indeed, but the bigger issue to me is still the ability to separate the initialization of the runtime and shutdown from the extension module.  The rest can be done relatively easily from the Python side.

> 
> > Is there a timeline for when numpy support will be available?
> 
> There's not an explicit timeline. We have some numpy support already (in that specific numpy arrays can be passed to Chapel functions today) but it isn't _full_ numpy support. We do have it on our priority list, but may not get back to extending it this release (and I expect full numpy support to take longer than that), it all depends on the other priorities and developer resources.

Oh, sorry I have not been writing any code to test, and I intend to at some point.  I am curious where I can learn more about the details of what numpy arrays are supported, the docs seem a bit scant on this or I am unable to find out.  Thank you!

**Comment by lydia-duncan** [comment]
> I am curious where I can learn more about the details of what numpy arrays are supported, the docs seem a bit scant on this or I am unable to find out.

Yeah, we ultimately decided not to explicitly list supported types, as that was liable to change as the feature became more stable and we didn't want to warn people against something unnecessarily if we'd added support for it and forgot to update the docs.

I'm a little rusty on the details myself, but I believe the numpy array must:
- be 1 dimensional
- be contiguous
- contain elements of types that can be passed to Chapel functions on their own (ints, uints, etc)
- start at index 0

And there may be other limitations I'm not remembering right now.  Also, note that the array will currently get copied into the Chapel function, so it will have to be returned to see modifications that have occurred to it.

**Comment by bradcray** [comment]
Thanks for filing this as a follow-up to our SciPy conversation, @prabhuramachandran.  As mentioned there, if there are aspects of this proposal that would increase the likelihood of you using Chapel (e.g., NumPy support for multidimensional arrays, different setup/teardown strategies), that would be valuable for us to know and could move things up in priority that are currently waiting for a champion in the user community.

**Comment by prabhuramachandran** [comment]
Hi @bradcray -- thanks for your comments at SciPy.  I have been slowly re-learning Chapel, I'd forgotten everything that I picked up a few years ago but it is slow going as I have too many things going on right now.  There are many things that I am concerned about though.  As I had mentioned above, the setup/teardown issue is certainly critical.  Some form of numpy support ideally with zero copy at least for contiguous arrays. Not supporting everything is fine (for example object arrays would be out of question) but some core subset would be great.  GPU support for sure although there are many things that are unclear there.  

I am not sure how zero copy will work or how arrays would work via gasnet.  So what happens to a numpy array if this is a distributed computation and not on a single machine. The general Python crowd will surely want scipy and pandas as well and I really have no idea how that will ever work or even if it should and have the same issues with our own work in compyle (which is a tiny fly as compared to chapel).  How would things like LAPACK support work when you add GPU support?  These seem rather hard problems.  Perhaps a conversation would be easier.  

For a start, the ability to use chapel like OpenCL or CUDA from python.  I.e. the clean setup/teardown and basic numpy support will be great.  This along with the GPU support is itself huge but then again I do not know how the GPU support will look in Chapel and how seamless that will be.

**Comment by bradcray** [comment]
> I am not sure how zero copy will work or how arrays would work via gasnet. So what happens to a numpy array if this is a distributed computation and not on a single machine.

The current philosophy we've been trying to follow for this is that if the Chapel code and Python code are running on the same compute node (say, my laptop), then it ought to be possible to share data between them without doing copies.  But if the Chapel code is running out on a cluster (say) and the Python is on a login node, then zero-copy won't be possible.  Since Chapel has argument intents like `in` for "copy-in" or `ref` for "pass-by-reference", the thought is that if a Chapel routine is exported whose argument intent isn't compatible with the execution mode, the compiler will complain and refuse to compile it.  Thus, if I were to export a Chapel routine that took a string or array by reference:  `export proc foo(ref s: string, ref A: [] int) { ... }` it should work and be zero-copy for Python or C code running on the same system (e.g., when CHPL_LAUNCHER=none and CHPL_COMM=none) but should give an error otherwise (e.g., when using GASNet).  In which case the user would presumably want/need to write `export proc foo(in s: string, in A: [] int) { ... }` to assert that a copy-in is required as necessitated by the need to copy across the network.  That said, I can't recall how much or little our current implementation reflects this philosophy, and think we're inconsistent between strings/bytes vs. arrays vs. other scalar types, IIRC.

> Some form of numpy support ideally with zero copy at least for contiguous arrays.

I totally agree.  This is another case where our desire to implement it is strong, but it hasn't been a priority for our users, so it's languished in favor of other things for the time being.

**Comment by prabhuramachandran** [comment]
@bradcray -- What you propose above sounds very reasonable.  Do you have a roadmap of sorts that folks can look at?  I understand if things are still in flux.

**Comment by bradcray** [comment]
Sorry for the slow reply (busy week).  We don't have a detailed public-facing roadmap, but you can get a sense of our current set of priorities by looking in the Chapel [release notes](https://chapel-lang.org/releaseNotes.html), particularly the ["proposed priorities" slides](https://chapel-lang.org/releaseNotes/1.21/07-priorities.pdf).  All that said, we use an agile process with a 2-week sprint, so if new requirements or compelling use cases come online, these can shift as needed.

**Comment by JoyMonteiro** [comment]
Hello,

I mostly second the requirements @prabhuramachandran has described. We have been hoping to move the compiled code underlying the python code to chapel in [climt](https://github.com/CliMT/climt), since this will solve issues we currently face with having to write openmp style cython code everytime a new component is added.

Our current flow is like this: we pass a 3-4d numpy array to cython, which then iterates over one of the dimensions and calls a fortran subroutine on each sub-array. My understanding is that wrapping fortran code for chapel is not an issue.

The one (and only one) thing we need is transparent handling of contiguous multidimensional python arrays (integer or float). Since we mostly target shared memory systems for now, this will immediately allow us to get rid of the cython wrappers and use chapel instead.

Another question that keeps coming to mind is how we can use Python in the distributed case. My understanding is that the Python process runs on only one node and the chapel runtime instantiates processes on all nodes. In this case, how do we pass arrays instantiated at the Python level to each process? with MPI, we have a Python instance on each node which can independently allocate arrays and talk only to code on the same node.

Another approach we are considering is to move everything to chapel with some sort of python "server" which talks to it (more or less like arkouda), but then we have to rewrite (or re-wrap) crucial libraries for units handling, labeled arrays, ML -- things that we get more or less free when using Python.

Thanks a lot for engaging with us Python users; I understand we are not representative of your typical use-cases.

**Comment by bradcray** [comment]
Hi @JoyMonteiro —

Thanks for characterizing use cases of Chapel-Python interop that could be of interest or use to you.  Another of our key users has indicated a desire for passing multidimensional NumPy arrays to Chapel and back again (and the sooner the better), so it's on my mental list of things to try and get into our next release (it's been an intention of ours for awhile now, and seems that it shouldn't be too hard, yet it hadn't risen to the near-top of my list until now...).

The current Python -> distributed Chapel use cases that we've been looking at take the approach of requiring the arguments to be passed by `in` intent to acknowledge that a copy is being made from an interactive login node to the locale / compute node handling the call.  But if you wanted to pass the data to every locale / node, that would currently need to be done within the Chapel code itself.  For example, an approach like the following ought to work if the goal is truly to give each locale its own complete copy of `x`:

```chapel
export proc callFromPython(in x: ...) {
  coforall loc in Locales do
    on loc {
      var myLocX = x;
    }
}
```

Another approach might be to pass in an array and then assign it to a distributed array if the goal were to have each compute node own a portion of `x`, but not the whole thing (realizing in saying this that if Python could store the whole thing, each compute node could probably _afford_ to do so as well; but using a distributed array could provide other benefits like supporting distributed `forall` loops over the array, e.g.).

We've also talked about Chapel-Python interop modse in which we run a Python process per compute node, but AFAIK, I don't think we've been presented with any killer use cases for this yet, nor had anyone prototype such a mode that I can recall.  So if you have a computational pattern that really wanted this and could sketch our for us what it would look like (how you'd want the control to flow and what data to be passed around), that could be good input for our discussions and design.

W.r.t. your Arkouda-like client-server model, note that one of Arkouda's strengths that I find interesting is that not all of the code is written in Chapel; only the things that are expected to benefit from being written in Chapel (or that just haven't been bothered to be ported over yet because the author was better at Python).  In several cases, Arkouda routines are written on the Python side—either natively, or just calling out to Chapel for some portions of the operation yet not others.  As an example, Arkouda's set operations (union, intersection, difference, etc.) were originally written in Python for expedience, calling over to Chapel ops to implement aspects of them, but doing the rest in Python.  And, as I understand it, this was fast enough to be reasonable even though it wasn't necessarily the most performant or space-efficient way to do the operations.  Then we had a summer intern this past summer who converted them over to native Chapel-side operations, which resulted in a nice improvement in speed and memory use.  I'm not sure whether this model would be useful in your case or not, but wanted to point it out as a means of incrementally making greater use of Chapel over time without having to (re)write everything in Chapel on day one.

> Thanks a lot for engaging with us Python users; I understand we are not representative of your typical use-cases.

It's definitely different than why we set out to create Chapel, but given Chapel's appeal to Python users, it makes sense to try and flesh this story out as much as possible.  So we appreciate your input and indications of what would be helpful very much, thanks.

**Comment by bradcray** [comment]
PS — While we're talking about Chapel and Python, I wanted to mention in passing that one of our Google Summer of Code students this past summer added Chapel support for protobuf, which seems to be a FAQ'd question when I speak to Python crowds.  So I thought I'd mention it here in case it was of interest to anyone browsing this topic.  See https://chapel-lang.org/docs/master/tools/protoc-gen-chpl/protoc-gen-chpl.html for more information.

**Comment by JoyMonteiro** [comment]
Thanks @bradcray. I look forward to the next release!

Our plans are to test the waters by getting SMP up using chapel. If we eventually see a need for distributed Python-Chapel interactions, I will write back with a more fleshed out design.

**Comment by bradcray** [comment]
Sounds good.  If/when you get something up and running in SMP mode, let us know.  It'd be great to have someone kick the tires once we implement some improved NumPy interop support before the next release to try it out before it goes "live."  See the last bullet on slide 2 of [this deck](https://chapel-lang.org/releaseNotes/1.23/06-priorities.pdf) for my publicly stated intention to get this done.
